### PR TITLE
Event handler payment

### PR DESCRIPTION
**Pull request by arek1-618**
Event handler payment

### CODE REVIEW

File: install/install_schema/data.sql [hunk 1]
```diff
@@ -4754,6 +4755,7 @@ insert  into `vtiger_field`(`tabid`,`fieldid`,`columnname`,`tablename`,`generate
...
 insert  into `vtiger_field`(`tabid`,`fieldid`,`columnname`,`tablename`,`generatedtype`,`uitype`,`fieldname`,`fieldlabel`,`readonly`,`presence`,`defaultvalue`,`maximumlength`,`sequence`,`block`,`displaytype`,`typeofdata`,`quickcreate`,`quickcreatesequence`,`info_type`,`masseditable`,`helpinfo`,`summaryfield`,`fieldparams`,`header_field`,`maxlengthtext`,`maxwidthcolumn`,`visible`) values (90,2818,'ssingleorders_method_payments','u_yf_ssingleorders',1,16,'ssingleorders_method_payments','FL_METHOD_PAYMENTS',0,2,'','255',16,284,1,'V~O',1,0,'BAS',1,'',0,'',NULL,0,0,0);
+insert  into `vtiger_field`(`tabid`,`fieldid`,`columnname`,`tablename`,`generatedtype`,`uitype`,`fieldname`,`fieldlabel`,`readonly`,`presence`,`defaultvalue`,`maximumlength`,`sequence`,`block`,`displaytype`,`typeofdata`,`quickcreate`,`quickcreatesequence`,`info_type`,`masseditable`,`helpinfo`,`summaryfield`,`fieldparams`,`header_field`,`maxlengthtext`,`maxwidthcolumn`,`visible`) values (90,2819,'ssingleorders_payment_status','u_yf_ssingleorders',1,16,'ssingleorders_payment_status','FL_PAYMENT_STATUS',1,2,'PLL_NOT_PAID',NULL,17,284,10,'V~O',1,0,'BAS',1,'',0,'',NULL,0,0,0);
```
Comment (rskrzypczak) [inline install/install_schema/data.sql]:
```suggestion
insert  into `vtiger_field`(`tabid`,`fieldid`,`columnname`,`tablename`,`generatedtype`,`uitype`,`fieldname`,`fieldlabel`,`readonly`,`presence`,`defaultvalue`,`maximumlength`,`sequence`,`block`,`displaytype`,`typeofdata`,`quickcreate`,`quickcreatesequence`,`info_type`,`masseditable`,`helpinfo`,`summaryfield`,`fieldparams`,`header_field`,`maxlengthtext`,`maxwidthcolumn`,`visible`) values (90,2819,'ssingleorders_payment_status','u_yf_ssingleorders',1,15,'ssingleorders_payment_status','FL_PAYMENT_STATUS',1,2,'PLL_NOT_PAID',NULL,17,284,10,'V~O',1,0,'BAS',1,'',0,'',NULL,0,0,0);
```
uitype 15

File: install/install_schema/data.sql [hunk 2]
```diff
@@ -17707,6 +17710,13 @@ insert  into `vtiger_ssingleorders_method_payments`(`ssingleorders_method_paymen
 insert  into `vtiger_ssingleorders_method_payments`(`ssingleorders_method_paymentsid`,`ssingleorders_method_payments`,`presence`,`sortorderid`) values (3,'PLL_TRANSFER',1,3);
 insert  into `vtiger_ssingleorders_method_payments`(`ssingleorders_method_paymentsid`,`ssingleorders_method_payments`,`presence`,`sortorderid`) values (4,'PLL_CASH_ON_DELIVERY',1,4);
 
+/*Data for the table `vtiger_ssingleorders_payment_status` */
+
+insert  into `vtiger_ssingleorders_payment_status`(`ssingleorders_payment_statusid`,`ssingleorders_payment_status`,`presence`,`sortorderid`) values (1,'PLL_NOT_PAID',1,1);
```
Comment (rskrzypczak) [inline install/install_schema/data.sql]: presence = 0

File: modules/PaymentsIn/handlers/PaymentsInHandler.php [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+		if ($this->canUpdatePaymentStatus($recordModel)) {
+			$ordersId = (int) $recordModel->get('ssingleordersid');
+			$orderRecordModel = \Vtiger_Record_Model::getInstanceById($ordersId, 'SSingleOrders');
+			$sumOfGross = (float) $orderRecordModel->get('sum_gross');
+			$sumOfPayments = $this->getSumOfPayments($ordersId);
+			if ($sumOfGross > $sumOfPayments) {
+				$paymentStatus = 'PLL_UNDERPAID';
+			} elseif ($sumOfGross < $sumOfPayments) {
+				$paymentStatus = 'PLL_OVERPAID';
+			} elseif (\App\Validator::floatIsEqual($sumOfGross, 0.0, 2)) {
```
Comment (rskrzypczak) [inline modules/PaymentsIn/handlers/PaymentsInHandler.php]: ?

File: modules/PaymentsIn/handlers/PaymentsInHandler.php [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	 *
+	 * @return bool
+	 */
+	private function canUpdatePaymentStatus(Vtiger_Record_Model $recordModel): bool
+	{
+		if ($recordModel->isEmpty('ssingleordersid')) {
+			return false;
+		}
+		$ordersId = (int) $recordModel->get('ssingleordersid');
+		if ((int) $recordModel->get('currency_id') !== \App\Record::getCurrencyIdFromInventory($ordersId, 'SSingleOrders')) {
```
Comment (rskrzypczak) [inline modules/PaymentsIn/handlers/PaymentsInHandler.php]: ?

File: modules/PaymentsIn/handlers/PaymentsInHandler.php [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+	/**
+	 * Get the sum of all payments.
+	 *
+	 * @param int $recordId
+	 *
+	 * @return float
+	 */
+	private function getSumOfPayments(int $recordId): float
+	{
+		return (float) (new \App\Db\Query())
```
Comment (TomaszQr) [inline modules/PaymentsIn/handlers/PaymentsInHandler.php]: You should based on relation model

File: install/install_schema/data.sql [hunk 6]
```diff
@@ -3003,6 +3003,7 @@ insert  into `vtiger_eventhandlers`(`eventhandler_id`,`event_name`,`handler_clas
 insert  into `vtiger_eventhandlers`(`eventhandler_id`,`event_name`,`handler_class`,`is_active`,`include_modules`,`exclude_modules`,`priority`,`owner_id`) values (73,'EntityBeforeSave','Vtiger_RecordStatusHistory_Handler',1,'HelpDesk','',5,0);
 insert  into `vtiger_eventhandlers`(`eventhandler_id`,`event_name`,`handler_class`,`is_active`,`include_modules`,`exclude_modules`,`priority`,`owner_id`) values (74,'EntityAfterSave','Vtiger_RecordStatusHistory_Handler',1,'HelpDesk','',5,0);
 insert  into `vtiger_eventhandlers`(`eventhandler_id`,`event_name`,`handler_class`,`is_active`,`include_modules`,`exclude_modules`,`priority`,`owner_id`) values (75,'EntityBeforeSave','Products_Calculations_Handler',1,'Products,Services','',5,14);
+insert  into `vtiger_eventhandlers`(`eventhandler_id`,`event_name`,`handler_class`,`is_active`,`include_modules`,`exclude_modules`,`priority`,`owner_id`) values (76,'EntityAfterSave','PaymentsIn_PaymentsInHandler_Handler',1,'PaymentsIn','',5,0);
```
Comment (rskrzypczak) [inline install/install_schema/data.sql]: ?

File: install/install_schema/data.sql [hunk 7]
```diff
@@ -4758,6 +4759,7 @@ insert  into `vtiger_field`(`tabid`,`fieldid`,`columnname`,`tablename`,`generate
...
 insert  into `vtiger_field`(`tabid`,`fieldid`,`columnname`,`tablename`,`generatedtype`,`uitype`,`fieldname`,`fieldlabel`,`readonly`,`presence`,`defaultvalue`,`maximumlength`,`sequence`,`block`,`displaytype`,`typeofdata`,`quickcreate`,`quickcreatesequence`,`info_type`,`masseditable`,`helpinfo`,`summaryfield`,`fieldparams`,`header_field`,`maxlengthtext`,`maxwidthcolumn`,`visible`) values (35,2819,'purchase','vtiger_service',1,360,'purchase','FL_COST',0,0,'',NULL,4,92,1,'V~O',2,3,'BAS',0,'',1,'',NULL,0,0,0);
+insert  into `vtiger_field`(`tabid`,`fieldid`,`columnname`,`tablename`,`generatedtype`,`uitype`,`fieldname`,`fieldlabel`,`readonly`,`presence`,`defaultvalue`,`maximumlength`,`sequence`,`block`,`displaytype`,`typeofdata`,`quickcreate`,`quickcreatesequence`,`info_type`,`masseditable`,`helpinfo`,`summaryfield`,`fieldparams`,`header_field`,`maxlengthtext`,`maxwidthcolumn`,`visible`) values (90,2820,'ssingleorders_payment_status','u_yf_ssingleorders',1,15,'ssingleorders_payment_status','FL_PAYMENT_STATUS',1,2,'PLL_NOT_PAID',NULL,17,284,10,'V~O',1,0,'BAS',1,'',0,'',NULL,0,0,0);
```
Comment (rskrzypczak) [inline install/install_schema/data.sql]: maximumlength ?

File: modules/PaymentsIn/models/Module.php [hunk 8]
```diff
@@ -0,0 +1,74 @@
...
+	 * Update payment status.
+	 *
+	 * @param int $ordersId
+	 *
+	 * @return void
+	 */
+	public static function updatePaymentStatus(int $ordersId)
+	{
+		$orderRecordModel = \Vtiger_Record_Model::getInstanceById($ordersId, 'SSingleOrders');
+		$orderRecordModel->set(
```
Comment (rskrzypczak) [inline modules/PaymentsIn/models/Module.php]: check if field ssingleorders_payment_status is active

File: modules/PaymentsIn/handlers/PaymentsInHandler.php [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+	/**
+	 * Checking if you can update the payment status.
+	 *
+	 * @param Vtiger_Record_Model $recordModel
+	 *
+	 * @return bool
+	 */
+	private function canUpdatePaymentStatus(Vtiger_Record_Model $recordModel): bool
+	{
+		$fieldModel = Vtiger_Module_Model::getInstance('SSingleOrders')->getFieldByColumn('ssingleorders_payment_status');
```
Comment (rskrzypczak) [inline modules/PaymentsIn/handlers/PaymentsInHandler.php]: ?

**Pull request merged**